File: engine/src/engine/scripts/smoke_test.py
"""Phase 2 smoke test for the course-agent engine.

Run with ``uv run smoke``. Hits every external service the engine
depends on, prints a ✓ or ✗ per service, and exits non-zero on any
failure. The goal is "the wires are plugged in", not "the business
logic works" — that's Phase 6.

Checks:
  - Supabase reachable (REST root answers)
  - OpenRouter completion via DeepSeek (matches the default in
    docs/phase2.md; admins can swap to another model from /settings
    in Phase 6+)
  - Voyage AI embeddings (voyage-3-large, 1024-dim sanity check)
  - Serper search (organic results returned)
  - Langfuse + Sentry are only configured-or-not booleans for now;
    real ping logic lands in Phase 9.

Pydantic-Settings will raise ``ValidationError`` at startup if any
required variable is missing — that error is the loud-fail behaviour
Phase 2's acceptance criteria asks for.
"""

from __future__ import annotations

import sys
from typing import Callable

import httpx
from pydantic import ValidationError

from engine.config import settings


# DeepSeek slug used by the smoke test — matches the
# DEFAULT_MODEL_ASSIGNMENTS in app/src/lib/mock/llm-models.ts.
# Phase 6 will read the active assignment from a settings table.
SMOKE_OPENROUTER_MODEL = "deepseek/deepseek-chat-v3.1"


def _check(name: str, fn: Callable[[], None]) -> bool:
    try:
        fn()
    except Exception as exc:  # noqa: BLE001 — we want every failure logged
        print(f"✗ {name}: {exc}", file=sys.stderr)
        return False
    print(f"✓ {name}")
    return True


def main() -> None:
    try:
        cfg = settings()
    except ValidationError as exc:
        print("✗ env validation failed:", file=sys.stderr)
        for err in exc.errors():
            loc = ".".join(str(p) for p in err["loc"])
            print(f"  - {loc}: {err['msg']}", file=sys.stderr)
        sys.exit(1)

    print("smoke-test (engine)\n")

    def supabase() -> None:
        # The PostgREST root returns 200 with the OpenAPI schema or 401
        # when no apikey is sent. Either means the instance answered;
        # any other status code is a real connection problem.
        r = httpx.get(
            f"{cfg.supabase_url}rest/v1/",
            headers={"apikey": cfg.supabase_service_role_key},
            timeout=15.0,
        )
        if r.status_code not in (200, 401, 404):
            raise RuntimeError(f"HTTP {r.status_code}")

    def openrouter() -> None:
        r = httpx.post(
            "https://openrouter.ai/api/v1/chat/completions",
            headers={"Authorization": f"Bearer {cfg.openrouter_api_key}"},
            json={
                "model": SMOKE_OPENROUTER_MODEL,
                "messages": [{"role": "user", "content": "ping"}],
                "max_tokens": 5,
            },
            timeout=30.0,
        )
        r.raise_for_status()
        body = r.json()
        if not body.get("choices") or not body["choices"][0]["message"].get("content"):
            raise RuntimeError("empty completion")

    def voyage() -> None:
        r = httpx.post(
            "https://api.voyageai.com/v1/embeddings",
            headers={"Authorization": f"Bearer {cfg.voyage_api_key}"},
            json={
                "input": "ping",
                "model": "voyage-3-large",
                "input_type": "document",
            },
            timeout=30.0,
        )
        r.raise_for_status()
        dim = len(r.json()["data"][0]["embedding"])
        if dim != 1024:
            raise RuntimeError(f"expected 1024 dims, got {dim} — wrong model?")

    def serper() -> None:
        r = httpx.post(
            "https://google.serper.dev/search",
            headers={"X-API-KEY": cfg.serper_api_key},
            json={"q": "corporate training"},
            timeout=30.0,
        )
        r.raise_for_status()
        organic = r.json().get("organic", [])
        if not organic:
            raise RuntimeError("no organic results returned")

    ok = True
    ok &= _check("Supabase reachable", supabase)
    ok &= _check(f"OpenRouter completion ({SMOKE_OPENROUTER_MODEL})", openrouter)
    ok &= _check("Voyage AI embedding (voyage-3-large, 1024-dim)", voyage)
    ok &= _check("Serper search", serper)

    print()
    print(
        f"Langfuse:  {'configured' if cfg.langfuse_public_key else 'not configured (optional)'}"
    )
    print(
        f"Sentry:    {'configured' if cfg.sentry_dsn else 'not configured (optional)'}"
    )

    print()
    if not ok:
        print("One or more checks failed — Phase 2 is not done yet.", file=sys.stderr)
        sys.exit(1)
    print("All checks passed.")


if __name__ == "__main__":
    main()

File: app/scripts/smoke-test.ts
/**
 * Phase 2 smoke test for the Next.js dashboard.
 *
 * Runs from the project root: `pnpm smoke` (see package.json). Hits every
 * external service the dashboard depends on and exits non-zero on any
 * failure. Prints a ✓ or ✗ per service; the goal is "the wires are
 * plugged in", not "the business logic works".
 *
 *   - Supabase reachable (REST root responds, 200 or 401 both count)
 *   - GAS email webhook accepts the shared secret (sends to a sink address)
 *   - GAS email webhook rejects requests *without* the secret (401 expected)
 *   - Slack webhook posts (if configured)
 *
 * The GAS sink address defaults to a Phase 2 mailbox you control. Override
 * with $SMOKE_SINK_EMAIL so reviewers' real inboxes never see test traffic.
 */

import { config as loadDotenv } from "dotenv";
import { env } from "../src/lib/env";

// Load .env.local before reading process.env. Falls back gracefully if absent
// so the missing-var error still fires from `env()`.
loadDotenv({ path: ".env.local", quiet: true });

const SINK = process.env.SMOKE_SINK_EMAIL ?? "course-agent-smoke@edstellar.com";

type Check = {
  name: string;
  required: boolean;
  run: () => Promise<void>;
};

async function postJson(
  url: string,
  body: unknown,
  init: RequestInit = {},
): Promise<Response> {
  return fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json", ...init.headers },
    body: JSON.stringify(body),
    ...init,
  });
}

function buildChecks(e: ReturnType<typeof env>): Check[] {
  const checks: Check[] = [
    {
      name: "Supabase reachable",
      required: true,
      async run() {
        const res = await fetch(`${e.SUPABASE_URL}/rest/v1/`, {
          headers: { apikey: e.NEXT_PUBLIC_SUPABASE_ANON_KEY },
        });
        // 200 (with prefer=resolution) or 401 (no auth path) both mean
        // the instance answered. Anything else is a real connection issue.
        if (![200, 401, 404].includes(res.status)) {
          throw new Error(`HTTP ${res.status}`);
        }
      },
    },
  ];

  if (e.GAS_EMAIL_WEBHOOK_URL && e.GAS_EMAIL_SHARED_SECRET) {
    checks.push(
      {
        name: "GAS email webhook accepts the shared secret",
        required: true,
        async run() {
          const res = await postJson(e.GAS_EMAIL_WEBHOOK_URL!, {
            to: SINK,
            subject: "[smoke] course-agent Phase 2 ping",
            html: "<b>If you can read this, the wire is plugged in.</b>",
            secret: e.GAS_EMAIL_SHARED_SECRET,
          });
          if (!res.ok) throw new Error(`HTTP ${res.status}`);
          const json = (await res.json().catch(() => null)) as
            | { ok?: boolean }
            | null;
          if (!json?.ok) throw new Error(`unexpected body: ${JSON.stringify(json)}`);
        },
      },
      {
        name: "GAS email webhook rejects requests without the secret",
        required: true,
        async run() {
          const res = await postJson(e.GAS_EMAIL_WEBHOOK_URL!, {
            to: SINK,
            subject: "should not arrive",
            html: "x",
            secret: "wrong-secret",
          });
          if (res.status !== 401) {
            throw new Error(
              `expected 401, got ${res.status} — the GAS doPost is not enforcing the shared secret`,
            );
          }
        },
      },
    );
  }

  if (e.SLACK_WEBHOOK_URL) {
    checks.push({
      name: "Slack webhook posts",
      required: false,
      async run() {
        const res = await postJson(e.SLACK_WEBHOOK_URL!, {
          text: ":wave: course-agent Phase 2 smoke test",
        });
        if (!res.ok) throw new Error(`HTTP ${res.status}`);
      },
    });
  }

  return checks;
}

function reportUnconfigured(e: ReturnType<typeof env>): void {
  const unconfigured: string[] = [];
  if (!e.GAS_EMAIL_WEBHOOK_URL || !e.GAS_EMAIL_SHARED_SECRET) {
    unconfigured.push("GAS email webhook (Phase 7)");
  }
  if (!e.SLACK_WEBHOOK_URL) unconfigured.push("Slack (optional)");
  if (!e.SENTRY_DSN) unconfigured.push("Sentry (optional)");
  if (unconfigured.length > 0) {
    console.log(`\nNot configured (skipped): ${unconfigured.join(", ")}`);
  }
}

async function main(): Promise<void> {
  let e: ReturnType<typeof env>;
  try {
    e = env();
  } catch (err) {
    console.error(`✗ env validation failed\n${(err as Error).message}`);
    process.exit(1);
  }

  console.log(`smoke-test (Next.js)  ·  sink=${SINK}\n`);
  const checks = buildChecks(e);
  let failed = 0;
  for (const c of checks) {
    try {
      await c.run();
      console.log(`✓ ${c.name}`);
    } catch (err) {
      failed += 1;
      const sev = c.required ? "✗" : "!";
      console.error(`${sev} ${c.name}: ${(err as Error).message}`);
    }
  }
  reportUnconfigured(e);
  console.log();
  if (failed > 0) {
    console.error(`${failed} check(s) failed — Phase 2 is not done yet.`);
    process.exit(1);
  }
  console.log("All configured checks passed.");
}

main().catch((err) => {
  console.error("unexpected error:", err);
  process.exit(1);
});
